diff --git a/.vscode/settings.json b/.vscode/settings.json
new file mode 100644
index 0000000..0f83523
--- /dev/null
+++ b/.vscode/settings.json
@@ -0,0 +1,6 @@
+{
+    "files.associations": {
+        "*.dump": "riscv",
+        "user.h": "c"
+    }
+}
\ No newline at end of file
diff --git a/Makefile b/Makefile
index b971308..1b861bb 100644
--- a/Makefile
+++ b/Makefile
@@ -43,7 +43,7 @@ endif
 
 # riscv64-unknown-elf- or riscv64-linux-gnu-
 # perhaps in /opt/riscv/bin
-#TOOLPREFIX = 
+TOOLPREFIX = riscv64-unknown-elf-
 
 # Try to infer the correct TOOLPREFIX if not set
 ifndef TOOLPREFIX
@@ -157,6 +157,9 @@ UPROGS=\
 	$U/_grind\
 	$U/_wc\
 	$U/_zombie\
+	$U/_sleep\
+	$U/_pingpong\
+	$U/_find\
 
 
 ifeq ($(LAB),syscall)
@@ -219,7 +222,8 @@ docker:
 	docker $(DOCKER_RUN_ARGS)
 
 # try to generate a unique GDB port
-GDBPORT = $(shell expr `id -u` % 5000 + 25000)
+# GDBPORT = $(shell expr `id -u` % 5000 + 25000)
+GDBPORT = 26000
 # QEMU's gdb stub command line changed in 0.11
 QEMUGDB = $(shell if $(QEMU) -help | grep -q '^-gdb'; \
 	then echo "-gdb tcp::$(GDBPORT)"; \
@@ -242,15 +246,11 @@ qemu-gdb: $K/kernel .gdbinit fs.img
 	@echo "*** Now run 'gdb' in another window." 1>&2
 	$(QEMU) $(QEMUOPTS) -S $(QEMUGDB)
 
-GDBARGS = 
-GDBARGS += -ex 'set architecture riscv:rv64'
-GDBARGS += -ex 'target remote 127.0.0.1:$(GDBPORT)'
-GDBARGS += -ex 'symbol-file kernel/kernel'
-GDBARGS += -ex 'set riscv use-compressed-breakpoints yes'
-
-
 gdb: 
-	$(GDB) $(GDBARGS)
+	$(GDB) \
+ 		-ex 'file $(K)/kernel' \
+ 		-ex 'set arch riscv:rv64' \
+# 		-ex 'target remote localhost:$(GDBPORT)'
 
 ##
 ##  FOR testing lab grading script
diff --git a/commit/commands.gdb b/commit/commands.gdb
new file mode 100644
index 0000000..f6325e9
--- /dev/null
+++ b/commit/commands.gdb
@@ -0,0 +1,6 @@
+advance exec.c:90
+p myproc()->name
+n
+p myproc()->name
+da
+exit
diff --git a/commit/report.md b/commit/report.md
new file mode 100644
index 0000000..04fe1b9
--- /dev/null
+++ b/commit/report.md
@@ -0,0 +1,29 @@
+## Q.A.
+### Question. 1
+- `argc`的值为3. `argv`的大小是3*8=24.
+- `argv[0]`指向程序的路径, `argv[1]`, `argv[2]`依次指向被压入启动栈帧的"hello"和"world\n".
+- `sys_write`, `sys_sleep`.
+  
+### Question. 2
+- 使用两个管道进行信息传递的做法是平凡的, 这里尝试使用一个管道. 首先, 创建一个管道. 接着, `fork`出一个子进程. 令子进程与父进程同时复用管道的读端和写端即可. 
+  
+    这里涉及到的一个问题在于, 父进程的流程是先写后读, 而子进程的流程是先读后写. 有必要在父进程完成写操作后转让控制权, 否则刚刚被写入的信息就会被父进程读取, 此后, 子进程就会卡死在`read`操作了. 我首先尝试使用`sleep(0)`插入父进程的读写操作之间, 但这并不凑效, 原因在于无法进入`sys_sleep`内部的循环, 也就不能触发`sched`调度. 在更改为`sleep(1)`后, 问题得到了解决.
+- 在wc程序看来, 管道是仍然有效的, 因此始终等待被管道读端替代的`stdin`继续输入. wc需要一个`EOF`信号来确认不再有更多输入才能继续执行, 而关闭管道写端之后, 系统会给wc发送一个`EOF`信号, 所以取消注释就能顺利执行.
+
+## 实验思路
+### sleep
+照抄.
+
+### pipe
+考虑到需要主动制造调度这一点就可以了.
+
+### find
+简单的DFS以及基本的字符串处理. 声明一个全局的`cur_path`数组作为栈即可.
+
+### xv6-boot
+通过六条指令解决了需求.
+
+基本上很快可以确定是在`exec()`处, `proc`数组的元数据(`name`字段)发生了改变.
+因此, 直接`advance`到`exec()`调用处, 先后各打印一次`myproc()->name`即可.
+
+顺带一提, 最开始我使用先断点后`continue`的方式, 发现无论如何都需要至少七条指令; 经查询手册后发现改为`advance`就OK啦.
\ No newline at end of file
diff --git a/commit/test b/commit/test
new file mode 100755
index 0000000..ea66495
Binary files /dev/null and b/commit/test differ
diff --git a/commit/test.c b/commit/test.c
new file mode 100644
index 0000000..5321f06
--- /dev/null
+++ b/commit/test.c
@@ -0,0 +1,23 @@
+#include <unistd.h>
+
+int
+main(void) {
+    int p[2];
+    char *argv[2];
+    argv[0] = "wc";
+    argv[1] = NULL;
+    pipe(p);
+
+    if (fork() == 0) {
+        close(STDIN_FILENO);
+        dup(p[0]);
+        close(p[0]);
+        // close(p[1]);
+        execv("/bin/wc", argv);
+    } else {
+        close(p[0]);
+        write(p[1], "hello world\n", 12);        
+        close(p[1]);
+    }
+    return 0;
+}
\ No newline at end of file
diff --git a/user/find.c b/user/find.c
new file mode 100644
index 0000000..dae0ff6
--- /dev/null
+++ b/user/find.c
@@ -0,0 +1,93 @@
+#include "kernel/types.h"
+#include "kernel/stat.h"
+#include "kernel/fs.h"
+#include "user.h"
+
+static char cur_path[512];
+static int top;
+
+char*
+bare_name(char* full_path) {
+    char* p = full_path + strlen(full_path);
+    while (p > full_path && *p != '/') p--;
+    return p + 1;
+}
+
+void
+push_dir(char* new_dir) {
+    // new_dir should be produced by bare_name()
+    cur_path[top++] = '/';
+    strcpy(cur_path + top, new_dir);
+    top += strlen(new_dir);
+    cur_path[top] = 0;
+}
+
+void
+pop_dir() {
+    if (top > 0) {
+        top--;
+        while (top > 0 && cur_path[top] != '/') top--;
+        cur_path[top] = 0;
+    } else {
+        printf("error");
+        exit(-1);
+    }
+}
+
+
+void
+find(char* name) {
+    int fd = open(cur_path, 0);
+    if (fd < 0) {
+        printf("find: failed to open %s\n", cur_path);
+        exit(-1);
+    }
+
+    struct dirent de;
+    struct stat st;
+    char buf[512];
+    int idx = strlen(cur_path);
+    strcpy(buf, cur_path);
+    buf[idx++] = '/';
+    
+    while (read(fd, &de, sizeof(de)) == sizeof(de)) {
+        if (strcmp(de.name, name) == 0) {
+            printf("%s/%s\n", cur_path, de.name);
+        }
+        if (de.inum == 0 || strcmp(de.name, ".") == 0 || strcmp(de.name, "..") == 0) {
+            continue;
+        }
+        strcpy(buf + idx, de.name);
+        buf[idx + strlen(de.name)] = 0;
+        if (stat(buf, &st) < 0) {
+            printf("find: failed to stat %s\n", buf);
+            continue;
+        }
+        if (st.type == T_DIR) {
+            push_dir(de.name);
+            find(name);
+            pop_dir();
+        } else {
+            // do nothing
+            ;
+        }
+    }
+
+    close(fd);
+}
+
+int
+main(int argc, char* argv[]) {
+    if (argc != 3) {
+        printf("usage: find <path> <name>\n");
+        exit(-1);
+    }
+
+    for (; *argv[1]; argv[1]++) {
+        cur_path[top++] = *argv[1];
+    }
+
+    find(argv[2]);
+
+    exit(0);
+}
\ No newline at end of file
diff --git a/user/pingpong.c b/user/pingpong.c
new file mode 100644
index 0000000..94c5cc1
--- /dev/null
+++ b/user/pingpong.c
@@ -0,0 +1,42 @@
+#include "kernel/types.h"
+#include "user.h"
+
+int
+main(void) {
+    int pid;
+    int p[2];
+    if (pipe(p) < 0) {
+        printf("pingpong: pipe failed\n");
+        exit(-1);
+    }
+    if ((pid = fork()) < 0) {
+        printf("pingpong: fork failed\n");
+        exit(-1);
+    } else if (pid == 0) {
+        // child
+        int parent_id = -1;
+        int child_id = getpid();
+        read(p[0], &parent_id, sizeof(int));
+        printf("%d: received ping from pid %d\n", child_id, parent_id);
+        write(p[1], &child_id, sizeof(int));
+        close(p[0]);
+        close(p[1]);
+        exit(0);
+    } else {
+        // parent
+        int parent_id = getpid();
+        int child_id = -1;
+        write(p[1], &parent_id, sizeof(int));
+
+        // key: cause a yield
+        sleep(1);
+        
+        read(p[0], &child_id, sizeof(int));
+        printf("%d: received pong from pid %d\n", parent_id, child_id);
+        int exit_code;
+        wait(&exit_code);
+        close(p[0]);
+        close(p[1]);
+    }
+    exit(0);
+}
\ No newline at end of file
diff --git a/user/sleep.c b/user/sleep.c
new file mode 100644
index 0000000..46ca5f7
--- /dev/null
+++ b/user/sleep.c
@@ -0,0 +1,13 @@
+#include "kernel/types.h"
+#include "user.h"
+
+int main(int argc, char* argv[]) {
+    if (argc != 2) {
+        printf("Sleep needs one argument!\n");
+        exit(-1);
+    }
+    int ticks = atoi(argv[1]);
+    sleep(ticks);
+    printf("(nothing happens for a little while)\n");
+    exit(0);
+}
\ No newline at end of file
